internal/shim-sgx/src/handler/mod.rs

```diff
@@ -34,29 +38,60 @@ use core::mem::size_of;
 use core::ptr::read_unaligned;
 use core::ptr::NonNull;
 
+use mmledger::Access;
+use primordial::{Address, Offset, Page};
 use sallyport::guest::Handler as _;
 use sallyport::guest::{self, Platform, ThreadLocalStorage};
 use sallyport::item::enarxcall::sgx::{Report, ReportData, TargetInfo, TECH};
 use sallyport::item::enarxcall::SYS_GETATT;
 use sallyport::item::syscall::{ARCH_GET_FS, ARCH_GET_GS, ARCH_SET_FS, ARCH_SET_GS};
-use sallyport::libc::{off_t, EINVAL, EMSGSIZE, ENOSYS, STDERR_FILENO};
-
-use crate::handler::usermem::UserMemScope;
-use crate::{shim_address, DEBUG, ENARX_EXEC_END, ENARX_EXEC_START, ENCL_SIZE};
+use sallyport::libc::{
+    off_t, EINVAL, EMSGSIZE, ENOMEM, ENOSYS, MAP_ANONYMOUS, MAP_PRIVATE, PROT_EXEC, PROT_READ,
+    PROT_WRITE, STDERR_FILENO,
+};
+use sgx::page::{Class, Flags};
 use sgx::ssa::StateSaveArea;
 use sgx::ssa::Vector;
+use spinning::{Lazy, RwLock};
+use x86_64::addr::VirtAddr;
+use x86_64::structures::paging::Page as PageAddr;
 
 // Opcode constants, details in Volume 2 of the Intel 64 and IA-32 Architectures Software
 // Developer's Manual
 const OP_SYSCALL: u16 = 0x050f;
 const OP_CPUID: u16 = 0xa20f;
 
+/// The keep heap
+pub static HEAP: Lazy<RwLock<Heap>> = Lazy::new(|| {
+    let start = unsafe { &ENARX_EXEC_END as *const _ } as usize;
+    let end = shim_address() + ENCL_SIZE;
+    RwLock::new(Heap::new(Address::new(start), Address::new(end)))
+});
+
+// For `Handler::prepare_mmap()`
+static ZERO: Page = Page::zeroed();
+
 /// Thread local storage for the current thread
 pub struct Handler<'a> {
     block: &'a mut [usize],
     ssa: &'a mut StateSaveArea,
 }
 
+impl<'a> Handler<'a> {
+    /// Acknowledge pages prepare_mmaped by the host.
+    fn prepare_mmap(&mut self, addr: Address<usize, Page>, length: Offset<usize, Page>) {
+        let secinfo = Class::Regular.info(Flags::READ | Flags::WRITE | Flags::EXECUTE);
+        let zero = PageAddr::from_start_address(VirtAddr::new(ZERO.as_ptr() as u64)).unwrap();
+        for i in 0..length.items() {
+            let virt_addr = VirtAddr::new((addr.raw() + i * Page::SIZE) as u64);
+            let page_addr = PageAddr::from_start_address(virt_addr).unwrap();
+            if secinfo.accept_copy(page_addr, zero).is_err() {
+                panic!();
+            }
```

Comment:
```suggestion
            secinfo.accept_copy(page_addr, zero).expect("accept_copy failed");
```

Reply:
or just `.unwrap()`

Reply:
Unfortunately cannot use because error type is not derived from `Debug`. I will fix this post-PR, once sgx crate has been fixed.

Reply:
Huh? `.expect()` or `.unwrap()` just panic the same way you do here.

Reply:
This is not resolved. Why did you close this?

Reply:
oops my fault, sorry... you are right, of course

Reply:
```
$ git diff
diff --git a/internal/shim-sgx/src/handler/mod.rs b/internal/shim-sgx/src/handler/mod.rs
index 2cabf2e..69a6aca 100644
--- a/internal/shim-sgx/src/handler/mod.rs
+++ b/internal/shim-sgx/src/handler/mod.rs
@@ -87,9 +87,7 @@ impl<'a> Handler<'a> {
         for i in 0..length.items() {
             let virt_addr = VirtAddr::new((region.start.raw() + i * Page::SIZE) as u64);
             let page_addr = PageAddr::from_start_address(virt_addr).unwrap();
-            if secinfo.accept_copy(page_addr, zero).is_err() {
-                panic!();
-            }
+            secinfo.accept_copy(page_addr, zero).expect("SecInfo::accept_copy()");
         }
     }
 }

mustikka in enarx/internal/shim-sgx on  feat/sgx/mmap [!?] is 📦 v0.3.0 via 🦀 v1.61.0-nightly 
$ cargo clippy
    Checking shim-sgx v0.3.0 (/home/jarkko/Projects/enarx/internal/shim-sgx)
error[E0277]: `sgx::page::sinfo::AcceptError` doesn't implement `core::fmt::Debug`
    --> src/handler/mod.rs:90:50
     |
90   |             secinfo.accept_copy(page_addr, zero).expect("SecInfo::accept_copy()");
     |                                                  ^^^^^^ `sgx::page::sinfo::AcceptError` cannot be formatted using `{:?}` because it doesn't implement `core::fmt::Debug`
     |
     = help: the trait `core::fmt::Debug` is not implemented for `sgx::page::sinfo::AcceptError`
note: required by a bound in `core::result::Result::<T, E>::expect`
    --> /home/jarkko/.rustup/toolchains/nightly-2022-03-23-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/result.rs:1030:12
     |
1030 |         E: fmt::Debug,
     |            ^^^^^^^^^^ required by this bound in `core::result::Result::<T, E>::expect`

For more information about this error, try `rustc --explain E0277`.
error: could not compile `shim-sgx` due to previous error
```

Reply:
> > ```
> > $ git diff
> > diff --git a/internal/shim-sgx/src/handler/mod.rs b/internal/shim-sgx/src/handler/mod.rs
> > index 2cabf2e..4c3fca8 100644
> > --- a/internal/shim-sgx/src/handler/mod.rs
> > +++ b/internal/shim-sgx/src/handler/mod.rs
> > @@ -87,9 +87,7 @@ impl<'a> Handler<'a> {
> >          for i in 0..length.items() {
> >              let virt_addr = VirtAddr::new((region.start.raw() + i * Page::SIZE) as u64);
> >              let page_addr = PageAddr::from_start_address(virt_addr).unwrap();
> > -            if secinfo.accept_copy(page_addr, zero).is_err() {
> > -                panic!();
> > -            }
> > +            secinfo.accept_copy(page_addr, zero).unwrap();
> >          }
> >      }
> >  }
> > 
> > mustikka in enarx/internal/shim-sgx on  feat/sgx/mmap [!?] is 📦 v0.3.0 via 🦀 v1.61.0-nightly 
> > $ cargo clippy
> >     Checking shim-sgx v0.3.0 (/home/jarkko/Projects/enarx/internal/shim-sgx)
> > error[E0277]: `sgx::page::sinfo::AcceptError` doesn't implement `core::fmt::Debug`
> >     --> src/handler/mod.rs:90:50
> >      |
> > 90   |             secinfo.accept_copy(page_addr, zero).unwrap();
> >      |                                                  ^^^^^^ `sgx::page::sinfo::AcceptError` cannot be formatted using `{:?}` because it doesn't implement `core::fmt::Debug`
> >      |
> >      = help: the trait `core::fmt::Debug` is not implemented for `sgx::page::sinfo::AcceptError`
> > note: required by a bound in `core::result::Result::<T, E>::unwrap`
> >     --> /home/jarkko/.rustup/toolchains/nightly-2022-03-23-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/result.rs:1073:12
> >      |
> > 1073 |         E: fmt::Debug,
> >      |            ^^^^^^^^^^ required by this bound in `core::result::Result::<T, E>::unwrap`
> > 
> > For more information about this error, try `rustc --explain E0277`.
> > error: could not compile `shim-sgx` due to previous error
> > ```
> 
> yeah, sorry, my fault.... apologies



> > ```
> > $ git diff
> > diff --git a/internal/shim-sgx/src/handler/mod.rs b/internal/shim-sgx/src/handler/mod.rs
> > index 2cabf2e..4c3fca8 100644
> > --- a/internal/shim-sgx/src/handler/mod.rs
> > +++ b/internal/shim-sgx/src/handler/mod.rs
> > @@ -87,9 +87,7 @@ impl<'a> Handler<'a> {
> >          for i in 0..length.items() {
> >              let virt_addr = VirtAddr::new((region.start.raw() + i * Page::SIZE) as u64);
> >              let page_addr = PageAddr::from_start_address(virt_addr).unwrap();
> > -            if secinfo.accept_copy(page_addr, zero).is_err() {
> > -                panic!();
> > -            }
> > +            secinfo.accept_copy(page_addr, zero).unwrap();
> >          }
> >      }
> >  }
> > 
> > mustikka in enarx/internal/shim-sgx on  feat/sgx/mmap [!?] is 📦 v0.3.0 via 🦀 v1.61.0-nightly 
> > $ cargo clippy
> >     Checking shim-sgx v0.3.0 (/home/jarkko/Projects/enarx/internal/shim-sgx)
> > error[E0277]: `sgx::page::sinfo::AcceptError` doesn't implement `core::fmt::Debug`
> >     --> src/handler/mod.rs:90:50
> >      |
> > 90   |             secinfo.accept_copy(page_addr, zero).unwrap();
> >      |                                                  ^^^^^^ `sgx::page::sinfo::AcceptError` cannot be formatted using `{:?}` because it doesn't implement `core::fmt::Debug`
> >      |
> >      = help: the trait `core::fmt::Debug` is not implemented for `sgx::page::sinfo::AcceptError`
> > note: required by a bound in `core::result::Result::<T, E>::unwrap`
> >     --> /home/jarkko/.rustup/toolchains/nightly-2022-03-23-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/result.rs:1073:12
> >      |
> > 1073 |         E: fmt::Debug,
> >      |            ^^^^^^^^^^ required by this bound in `core::result::Result::<T, E>::unwrap`
> > 
> > For more information about this error, try `rustc --explain E0277`.
> > error: could not compile `shim-sgx` due to previous error
> > ```
> 
> yeah, sorry, my fault.... apologies

I added a FIXME comment.